interfaces/builtin/dotfiles.go

```diff
@@ -0,0 +1,159 @@
+}
+
+func validatePaths(attrName string, paths []interface{}) error {
+	for _, npp := range paths {
+		np, ok := npp.(string)
+		if !ok {
+			return fmt.Errorf("%q must be a list of strings", attrName)
+		}
+		if strings.HasSuffix(np, "/") {
+			return fmt.Errorf(`%q cannot end with "/"`, np)
+		}
+		if !strings.HasPrefix(np, "/") && !strings.HasPrefix(np, "$HOME/") {
+			return fmt.Errorf(`%q must start with "/" or "$HOME"`, np)
+		}
+		if !strings.HasPrefix(np, "$HOME/") && strings.Contains(np, "$HOME") {
+			return fmt.Errorf(`$HOME must only be used at the start of the path of %q`, np)
+		}
+		if strings.Contains(np, "@{") {
+			return fmt.Errorf(`%q should not use "@{"`, np)
+		}
+		p := filepath.Clean(np)
+		if p != np {
+			return fmt.Errorf("%q must be clean", np)
+		}
+		if strings.Contains(p, "~") {
+			return fmt.Errorf(`%q contains invalid "~"`, p)
+		}
+		if err := apparmor.ValidateFreeFromAARE(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (iface *dotfilesInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	hasValidAttr := false
+	for _, att := range []string{"read", "write"} {
+		if _, ok := plug.Attrs[att]; !ok {
+			continue
+		}
+		il, ok := plug.Attrs[att].([]interface{})
+		if !ok {
+			return fmt.Errorf("cannot add dotfiles plug: %q must be a list of strings", att)
+		}
+		if err := validatePaths(att, il); err != nil {
+			return fmt.Errorf("cannot add dotfiles plug: %s", err)
+		}
+		hasValidAttr = true
+	}
+	if !hasValidAttr {
+		return fmt.Errorf(`cannot add dotfiles plug: needs valid "read" or "write" attribute`)
+	}
+
+	return nil
+}
+
+func formatPath(ip interface{}) (string, error) {
+	p, ok := ip.(string)
+	if !ok {
+		return "", fmt.Errorf("%[1]v (%[1]T) is not a string", ip)
+	}
+	prefix := ""
+	if strings.Count(p, "$HOME") > 0 {
+		p = strings.Replace(p, "$HOME", "@{HOME}", 1)
+		prefix = "owner "
+	}
+	p += "{,/,/**}"
+
+	return fmt.Sprintf("%s%q", prefix, filepath.Clean(p)), nil
+}
+
+func allowPathAccess(buf *bytes.Buffer, perm string, paths []interface{}) error {
+	for _, rawPath := range paths {
+		p, err := formatPath(rawPath)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(buf, "%s %s\n", p, perm)
+	}
+	return nil
+}
+
+func (iface *dotfilesInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var reads, writes []interface{}
+	_ = plug.Attr("read", &reads)
+	_ = plug.Attr("write", &writes)
+
+	errPrefix := fmt.Sprintf(`cannot connect plug %s: `, plug.Name())
+	buf := bytes.NewBufferString(dotfilesConnectedPlugAppArmor)
+	if err := allowPathAccess(buf, "rk,", reads); err != nil {
+		return fmt.Errorf("%s%v", errPrefix, err)
+	}
+	if err := allowPathAccess(buf, "rwkl,", writes); err != nil {
+		return fmt.Errorf("%s%v", errPrefix, err)
+	}
+	spec.AddSnippet(buf.String())
+
+	return nil
+}
+
+func init() {
+	registerIface(&dotfilesInterface{commonInterface{
+		name:                 "dotfiles",
```

Comment:
The suggested _files_ name feels a bit awkward in the sense the snap definitely has access to files.. many files. :)

I suggest naming the interface as _sensitive-files._ This remains general, while also giving the feeling that is restricted for good reasons.

Reply:
But is unclear that the file are "sensitive", it depends,  they are apriori inaccessible because we start from a conservative position. I think the trouble we are having with naming this is that is an expedient interface but problematic, it's hard for the name to capture what it allows because in this case it really fully depends on the attributes on it. We need to convey something in the range of "specific" to "otherwise-inaccessible" to indeed "sensitive".

Reply:
So thinking some more:

* "raw-files" or "bare-files",  it undescores that this give *direct* access to files *without any mediation or predefined purpose* (and that would not be accessible otherwise)
* "fenced-files" or something similar (fenced seemed the shortest word with this possible meaning), it underscores that the files would *not* be apriori *accessible*
* "sensitive-files",  it underscores that the file might be *sensitive* (that might be misleading in general) and were not previously accessible
* "designated-files" or "specific-files",  it would underscore mainly that this lets specify precise files/dirs

Reply:
Yeah, @pedronis is right. The interface will allow access to anything the developer chooses-- it's our policies around it and what the developer chooses that make it sensitive or not. Eg, ~/.config/foo might be sensitive, but /etc/foo might not. I don't hate "sensitive-files" but it doesn't seem quite right either.

I don't really like "raw-files" or "bare-files" for the same reason @niemeyer doesn't like "files"-- they are just files and no more raw or bare than files already allowed by the existing interface policy. "fenced-files" seems weird. "designated-files" and "specific-files" seems redundant and don't hit my ear quite right.

What about "extra-files"? It could be argued as redundant too, but at least it conveys that these are files separate from what would normally be allowed.

Reply:
Stepping back a bit, we have "content",  which we didn't call "shared-content" or "protected-content" or "private-content",

ATM I could live with "files" (tough generic/vacuous as it might be) or "misc-files"  (as in miscellaneous) which is similar to "extra-files"  but "extra" has a nuance of "(from) another place/set" somehow for me that seems confusing here.